sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,776 @@
+                            if group.is_joined:
+                                try:
+                                    self._join_group_core(group_name)
+                                except Exception as e:  # pylint: disable=broad-except
+                                    self._call_back(
+                                        CallBackType.REJOIN_GROUP_FAILED,
+                                        OnRejoinGroupFailedArgs(group=group_name, error=e),
+                                    )
+
+                    self._call_back(
+                        CallBackType.CONNECTED,
+                        OnConnectedArgs(connection_id=message.connection_id, user_id=message.user_id),
+                    )
+
+            def handle_disconnected_message(message: DisconnectedMessage):
+                self._last_disconnected_message = message
+
+            def handle_group_data_message(message: GroupDataMessage):
+                if message.sequence_id:
+                    if not self._sequence_id.try_update(message.sequence_id):
+                        # // drop duplicated message
+                        return
+
+                self._call_back(
+                    CallBackType.GROUP_MESSAGE,
+                    OnGroupDataMessageArgs(
+                        data_type=message.data_type,
+                        data=message.data,
+                        group=message.group,
+                        from_user_id=message.from_user_id,
+                        sequence_id=message.sequence_id,
+                    ),
+                )
+
+            def handle_server_data_message(message: ServerDataMessage):
+                if message.sequence_id:
+                    if not self._sequence_id.try_update(message.sequence_id):
+                        # // drop duplicated message
+                        return
+
+                self._call_back(
+                    CallBackType.SERVER_MESSAGE,
+                    OnServerDataMessageArgs(
+                        data_type=message.data_type, data=message.data, sequence_id=message.sequence_id
+                    ),
+                )
+
+            parsed_message = self._protocol.parse_messages(data)
+            if parsed_message is None:
+                #  None means the message is not recognized.
+                return
+            if parsed_message.kind == "connected":
+                handle_connected_message(parsed_message)
+            elif parsed_message.kind == "disconnected":
+                handle_disconnected_message(parsed_message)
+            elif parsed_message.kind == "ack":
+                handle_ack_message(parsed_message)
+            elif parsed_message.kind == "groupData":
+                handle_group_data_message(parsed_message)
+            elif parsed_message.kind == "serverData":
+                handle_server_data_message(parsed_message)
+            else:
+                _LOGGER.warning("unknown message type: {}", parsed_message.kind)
```

Comment:
Youll want to change the this to the %s style of formatting for loggers, so that its evaluation can be [deferred ](https://docs.python.org/3/howto/logging.html#optimization)

Reply:
updated